src/future/shared.rs

```diff
@@ -38,64 +39,84 @@ impl<F> fmt::Debug for Shared<F>
 {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
         fmt.debug_struct("Shared")
-            .field("id", &self.id)
             .field("inner", &self.inner)
+            .field("waiter", &self.waiter)
             .finish()
     }
 }
 
 struct Inner<F: Future> {
-    next_clone_id: Mutex<u64>,
-    state: Mutex<State<F>>,
+    next_clone_id: AtomicUsize,
+    future: UnsafeCell<Spawn<F>>,
```

Comment:
Ideally, we should drop the wrapped future as soon as it completes, to free any resources that it holds. The representation here does not allow that to happen. Maybe it should be `UnsafeCell<Option<Spawn<F>>>`? Or, even better, maybe we could pack the `future` and `result` fields into a single enum field. Would that work?